Привет, Виталий! 

Меня зовут Светлана Медведева и я буду проверять Твою работу. Предлагаю общаться на "ты". 

Просьба при доработке работы оставлять мои комментарии без изменений.

Комментарии я разделяю на следующие категории:

<div class="alert alert-block alert-success">
В случае если всё верно!
</div>


<div class="alert alert-block alert-warning">
В случае если можно что-то доработать, но эта доработка не критична или если есть варианты улучшения работы.
</div>

<div class="alert alert-block alert-danger">
Замечания, которые нужно исправить. Без исправления этих замечаний проект принят не может быть.
</div>

### <font color='blue'>Рецензия на работу</font>
* Отличная работа!
* Данные детально проанализированы и выполнена корректная замена пропусков в данных.
* Выявлены закономерности в просрочках кредитов.
* Поздравляю работа засчитана :)
* Успехов на следующих спринтах!

# Исследование надёжности заёмщиков  

Что такое проект  

Поздравляем! Вы прошли курс в тренажёре. Самое время проверить полученные знания на практике и решить аналитический кейс. Выполнять работу вы будете самостоятельно.
Как закончите работу над проектом, отправьте его на проверку ревьюеру. В течение суток вы получите комментарии. Их нужно учесть: доработать проект и вернуть ревьюеру обновлённый вариант.
Скорее всего, вы снова получите комментарии по кейсу. Это нормально — доработка может проходить в несколько этапов.
Проект завершён, когда засчитаны все исправления.  

Описание проекта  

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.
Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.  

Инструкция по выполнению  

### Шаг 1. Откройте таблицу и изучите общую информацию о данных  
Путь к файлу: /datasets/data.csv. Скачать датасет  

### Шаг 2. Предобработка данных  
1.	определите и заполните пропущенные значения:  
    * опишите, какие пропущенные значения вы обнаружили;  
    * приведите возможные причины появления пропусков в данных;  
    * объясните, по какому принципу заполнены пропуски;  
2.	замените вещественный тип данных на целочисленный:   
    * поясните, как выбирали метод для изменения типа данных;  
3.	удалите дубликаты:  
    * поясните, как выбирали метод для поиска и удаления дубликатов в данных;  
    * приведите возможные причины появления дубликатов;  
4.	выделите леммы в значениях столбца с целями получения кредита:   
    * опишите, как вы проводили лемматизацию целей кредита;  
5.	категоризируйте данные:   
    * перечислите, какие «словари» вы выделили для этого набора данных, и объясните, почему.  

В данных могут встречаться артефакты — значения, которые не отражают действительность. Например, отрицательное количество дней трудового стажа. Для реальных данных — это нормально. Нужно описать возможные причины появления таких данных и обработать их.  

### Шаг 3. Ответьте на вопросы   
•	Есть ли зависимость между наличием детей и возвратом кредита в срок?  
•	Есть ли зависимость между семейным положением и возвратом кредита в срок?  
•	Есть ли зависимость между уровнем дохода и возвратом кредита в срок?  
•	Как разные цели кредита влияют на его возврат в срок?   
Ответы сопроводите интерпретацией — поясните, о чём именно говорит полученный вами результат.   

### Шаг 4. Напишите общий вывод   
Оформление: Задание выполните в Jupyter Notebook. Программный код заполните в ячейках типа code, текстовые пояснения — в ячейках типа markdown. Примените форматирование и заголовки.  

Описание данных   
•	children — кол-ичество детей в семье  
•	days_employed — общий трудовой стаж в днях  
•	dob_years — возраст клиента в годах  
•	education — уровень образования клиента  
•	education_id — идентификатор уровня образования  
•	family_status — семейное положение  
•	family_status_id — идентификатор семейного положения  
•	gender — пол клиента  
•	income_type — тип занятости  
•	debt — имел ли задолженность по возврату кредитов  
•	total_income — ежемесячный доход  
•	purpose — цель получения кредита   

**Как будут проверять мой проект?**
На что обращают внимание при проверке проектов:  
•	Как вы описываете найденные в данных проблемы?  
•	Какие методы замены типов данных, обработки пропусков и дубликатов применяете?  
•	Умеете лемматизировать?  
•	Категоризируете данные? Почему именно таким образом?  
•	Выводите ли финальные данные в сводных таблицах?  
•	Применяете ли конструкцию try-except для обработки потенциальных ошибок?  
•	Соблюдаете ли структуру проекта и поддерживаете аккуратность кода?  
•	Какие выводы делаете?  
•	Оставляете ли комментарии к шагам?    
Держите под рукой шпаргалки и конспекты прошлых уроков.  
Успехов!


<div class="alert alert-block alert-warning">
Комментарий ревьюера v. 1:
    
Молодец, что кратко описал задачи проекта, но было бы здорово добавить содержание проекта со ссылками.
</div>

# Шаг 1. Откройте файл с данными и изучите общую информацию. 

In [1]:
import pandas as pd
import numpy as np

### **Ссылка на Файлы всех проектов - Описание, НоутБуки+Датасеты и др. на Google-диске**

https://drive.google.com/drive/folders/1JCBOKeIEPSfMKowXS8qAgz_T9D4difg4?usp=sharing

In [2]:
data = pd.read_csv('/datasets/data.csv')

In [3]:
print(data.shape)
print()
print(data.info())

(21525, 12)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB
None


Типы данных в принципе можно было бы не менять, но по заданью требуется float заменить на int - в столбце total_income и days_employed. И дополнительно поменяем в столбце debt тип заменим на bool. Так будет наглядней и эффективней.

<div class="alert alert-block alert-success">
Комментарий ревьюера v. 1:
    
Да, верно у некоторых колонок нужно поменять тип данных.
</div>

In [4]:
print(f'Процент отсутствующих значений = {(21525-19351)/21525:,.2%}')

Процент отсутствующих значений = 10.10%


10,1% - это много, надо будет исправить.

<div class="alert alert-block alert-success">
Комментарий ревьюера v. 1:
    
Ты абсолютно верно заметил, что процент пропущенных данных слишком велик для удаления.
</div>

In [5]:
print(data.head(15))
print()
print(data.describe())

    children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   

In [6]:
# Вывод по 10 строк data
# a=10
# b=20
# key='n'
# while key == 'n':
#     print(data[a:b])
#     a+=10
#     b+=a+10
#     key = input()
#     if key=='b':
#         break

Все названия столбцов в нормальном формате (можно не переименовывать). Столбцы - days_employed и total_income имеют одинаковые кол-во non-null, здесь есть взаимосвязь. Возможно это те у кого еще нет стажа работы (молодые) и соответственно они еще не зарабатывают. Надо будет выяснить их возраст. 

<div class="alert alert-block alert-success">
Комментарий ревьюера v. 1:
    
Молодец, что заметил взаимосвязь пропусков в столбцах.
</div>

In [7]:
for column in data.columns:
    print(f"{column} : Пропуски/Пустые строки NaN = {data[column].isna().sum()}")

children : Пропуски/Пустые строки NaN = 0
days_employed : Пропуски/Пустые строки NaN = 2174
dob_years : Пропуски/Пустые строки NaN = 0
education : Пропуски/Пустые строки NaN = 0
education_id : Пропуски/Пустые строки NaN = 0
family_status : Пропуски/Пустые строки NaN = 0
family_status_id : Пропуски/Пустые строки NaN = 0
gender : Пропуски/Пустые строки NaN = 0
income_type : Пропуски/Пустые строки NaN = 0
debt : Пропуски/Пустые строки NaN = 0
total_income : Пропуски/Пустые строки NaN = 2174
purpose : Пропуски/Пустые строки NaN = 0


Nan - в 2 столбцах (days_employed, total_income) по 2174 значения

### Проверим каждый столбец:

In [8]:
for column in data.columns:
    print(data[column].value_counts())

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64
-986.927316     1
-7026.359174    1
-4236.274243    1
-6620.396473    1
-1238.560080    1
               ..
-2849.351119    1
-5619.328204    1
-448.829898     1
-1687.038672    1
-582.538413     1
Name: days_employed, Length: 19351, dtype: int64
35    617
40    609
41    607
34    603
38    598
42    597
33    581
39    573
31    560
36    555
44    547
29    545
30    540
48    538
37    537
50    514
43    513
32    510
49    508
28    503
45    497
27    493
56    487
52    484
47    480
54    479
46    475
58    461
57    460
53    459
51    448
59    444
55    443
26    408
60    377
25    357
61    355
62    352
63    269
64    265
24    264
23    254
65    194
66    183
22    183
67    167
21    111
0     101
68     99
69     85
70     65
71     58
20     51
72     33
19     14
73      8
74      6
75      1
Name: dob_years, dtype: int64
среднее   

### 1 Столбец 'children'

In [9]:
print(data['children'].value_counts())

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64


Больше всего бездетных, затем 1 и 2 ребенка (думамаю это логично). Плохие значения "20 - их 76" и "-1 их 47" здесь есть закономерность (они попали сюда одним и тем же способом), т.к. маловероятно чтобы все 76 и 47 строк принимали одно и тоже значение именно  20 и -1 (иначе были бы ошибки с другими цифрами типа 30,40,-4,-2). Проверить в каких строках оно встречается, и если на результат не будет оказывать влияния, то можно их удалить (их незначительное кол-во 123 против 21525 - меньше 0,6%)

<div class="alert alert-block alert-success">
Комментарий ревьюера v. 1:
    
Отличный анализ :)
</div>

### 2 Столбец 'days_employed'

In [10]:
print(data['days_employed'])
print()
print("Отрицательные значения = ",data[data['days_employed']<0]['days_employed'].count())
print()
print(data['days_employed'].describe())

0         -8437.673028
1         -4024.803754
2         -5623.422610
3         -4124.747207
4        340266.072047
             ...      
21520     -4529.316663
21521    343937.404131
21522     -2113.346888
21523     -3112.481705
21524     -1984.507589
Name: days_employed, Length: 21525, dtype: float64

Отрицательные значения =  15906

count     19351.000000
mean      63046.497661
std      140827.311974
min      -18388.949901
25%       -2747.423625
50%       -1203.369529
75%        -291.095954
max      401755.400475
Name: days_employed, dtype: float64


Присутствуют слишком большие значения (возможно это стаж в часах, а не днях) и много отрицательных  (возможно минус это тире) - оба вида недопустимы.
Если будет время запросить у источника дополнительную информацию и внести исправления и проанализировать закономерности в некоррктных значениях. На данный момент удалим чтобы не мешал, так как при выполнении поставленной задачи, эта информация не нужна

<div class="alert alert-block alert-success">
Комментарий ревьюера v. 1:
    
Ты верно отметил, что желательно запросит дополнительную информацию у заказчика/разработчиков.
</div>

### 3 Столбец 'dob_years'

In [11]:
print(data['dob_years'].value_counts().sort_index())

0     101
19     14
20     51
21    111
22    183
23    254
24    264
25    357
26    408
27    493
28    503
29    545
30    540
31    560
32    510
33    581
34    603
35    617
36    555
37    537
38    598
39    573
40    609
41    607
42    597
43    513
44    547
45    497
46    475
47    480
48    538
49    508
50    514
51    448
52    484
53    459
54    479
55    443
56    487
57    460
58    461
59    444
60    377
61    355
62    352
63    269
64    265
65    194
66    183
67    167
68     99
69     85
70     65
71     58
72     33
73      8
74      6
75      1
Name: dob_years, dtype: int64


"0" - выбивающееся значение и из общей тенденции. Проверить в каких строках оно встречается, и если на результат не будет оказывать влияния, то можно их удалить (их незначительное кол-во 101 против 21525 - меньше 0,5%)

<div class="alert alert-block alert-success">
Комментарий ревьюера v. 1:
    
Да, процент небольшой можем действительно удалить.
</div>

### 4 Столбец 'education' и 5 Столбец 'education_id '

In [12]:
print(data['education'].value_counts())
print(data['education_id'].value_counts())

среднее                13750
высшее                  4718
СРЕДНЕЕ                  772
Среднее                  711
неоконченное высшее      668
ВЫСШЕЕ                   274
Высшее                   268
начальное                250
Неоконченное высшее       47
НЕОКОНЧЕННОЕ ВЫСШЕЕ       29
НАЧАЛЬНОЕ                 17
Начальное                 15
ученая степень             4
Ученая степень             1
УЧЕНАЯ СТЕПЕНЬ             1
Name: education, dtype: int64
1    15233
0     5260
2      744
3      282
4        6
Name: education_id, dtype: int64


<div class="alert alert-block alert-success">
Комментарий ревьюера v. 1:
    
Верно, нужно удалить повторения.
</div>

### 6 Столбец 'family_status' и 7 Столбец 'family_status_id'

In [13]:
print(data['family_status'].value_counts())
print(data['family_status_id'].value_counts())

женат / замужем          12380
гражданский брак          4177
Не женат / не замужем     2813
в разводе                 1195
вдовец / вдова             960
Name: family_status, dtype: int64
0    12380
1     4177
4     2813
3     1195
2      960
Name: family_status_id, dtype: int64


Количество совпадает. Оба столбца корректны.

### 8 Столбец 'gender '

In [14]:
print(data['gender'].value_counts())

F      14236
M       7288
XNA        1
Name: gender, dtype: int64


Всего только 1 строка с некоррктным значением. Исправим его на "F"-женщина (так как видно, что женщин в 2 раза больше, то и вероятность того, что это женщина больше. К тому же для выполнения поставленной задачи столбец мало важен. 

<div class="alert alert-block alert-success">
Комментарий ревьюера v. 1:
    
Логичная замена :)
</div>

### 9 Столбец 'income_type'

In [15]:
print(data['income_type'].value_counts())

сотрудник          11119
компаньон           5085
пенсионер           3856
госслужащий         1459
безработный            2
предприниматель        2
в декрете              1
студент                1
Name: income_type, dtype: int64


Со столбцом все в порядке, оставим как есть.

### 10 Столбец 'debt'

In [16]:
print(data['debt'].value_counts())

0    19784
1     1741
Name: debt, dtype: int64


Все корректно

### 11 Столбец 'total_income'

In [17]:
print(data['total_income'].describe()) 
print(data[data['total_income'].isnull()].head(10))

count    1.935100e+04
mean     1.674223e+05
std      1.029716e+05
min      2.066726e+04
25%      1.030532e+05
50%      1.450179e+05
75%      2.034351e+05
max      2.265604e+06
Name: total_income, dtype: float64
    children  days_employed  dob_years education  education_id  \
12         0            NaN         65   среднее             1   
26         0            NaN         41   среднее             1   
29         0            NaN         63   среднее             1   
41         0            NaN         50   среднее             1   
55         0            NaN         54   среднее             1   
65         0            NaN         21   среднее             1   
67         0            NaN         52    высшее             0   
72         1            NaN         32    высшее             0   
82         2            NaN         50    высшее             0   
83         0            NaN         52   среднее             1   

            family_status  family_status_id gender  income_typ

In [18]:
# print(data['days_employed'].isnull().sum() & data['income_type'].isnull().sum())
# print(data.loc[data['days_employed'].isna()==True,'days_employed'].sum())

Похоже что нет закономерности в появлении пропусков с другими столбцами т.к. везде разные значения. Над этим столбцом надо поработать, так как он важен для выполненния задачи.

### 12 Столбец 'purpose'

In [19]:
print(data['purpose'].value_counts())

свадьба                                   797
на проведение свадьбы                     777
сыграть свадьбу                           774
операции с недвижимостью                  676
покупка коммерческой недвижимости         664
покупка жилья для сдачи                   653
операции с жильем                         653
операции с коммерческой недвижимостью     651
покупка жилья                             647
жилье                                     647
покупка жилья для семьи                   641
строительство собственной недвижимости    635
недвижимость                              634
операции со своей недвижимостью           630
строительство жилой недвижимости          626
покупка недвижимости                      624
покупка своего жилья                      620
строительство недвижимости                620
ремонт жилью                              612
покупка жилой недвижимости                607
на покупку своего автомобиля              505
заняться высшим образованием      

Столбец корректный, много совпадающих целей кредита - нужно совпадающие привезти к одному названию ( лемматизация )

### Вывод


В общем не так плохо: 

    1 children 21525 non-null int64 ---------------->!- Плохие значения "20" - их 76 (20 - очень мало вероятно) и "-1" их 47(нереально). Проанализировать. По-скольку их кол-во незначительно(76+47 против 21525) возможно удалить эти строки. 

    2 days_employed       19351 non-null float64 ---> - Есть проблемы: 1.Пропущениые, 2.Отрицательные, 3.Чрезмерно большие значения. Поскольку по условию задачи надобности в этом столбце нет, то возможно удалить его

    3 dob_years           21525 non-null int64 -----> -c 101 некорректное значение. Проверить в каких строках оно встречается, и если на результат не будет оказывать влияния, то можно их удалить (их незначительное кол-во 101 против 21525 - меньше 0,5%). Подходит для категоризации.

    4 education           21525 non-null object ----> -vd Привести все к одному фомату - сделать все буквы в нижнем регистре. сравнить совпадают ли кол-во значений по категориям со столбцом "education_id"

    5 education_id        21525 non-null int64 -----> + Корректный столбец - оставить как есть (сравнить совпадают ли кол-во значений по категориям со столбцом "education") 

    6 family_status       21525 non-null object ---->!+ Столбец корректен, оставить как есть

    7 family_status_id    21525 non-null int64 ----->!+ Столбец корректен, оставить как есть

    8 gender              21525 non-null object ----> - Исправим единственное значение "XNA" на "F"-женщина

    9 income_type         21525 non-null object ----> + Столбец корректен, оставить как есть

    10 debt                21525 non-null int64 ---->!+t Столбец корректен, а тип поменяем на bool

    11 total_income        19351 non-null float64 -->!-tc Процент отсутствующих значений = 10%. Слишком велик, чтобы пренебрегать. Надо чем-то заполнить. Также по заданью здесь надо поменять тип на int. Подходит для категоризации.

    12 purpose             21525 non-null object --->!+l Столбец корректный, много совпадающих целей кредита - нужно совпадающие привезти к одному названию ( лемматизация )
    
    ! - важные столбцы
    + - не требует обработки
    - - требуется обработка
    v - сверка корректности с логически родственным столбцом
    t - Поменять тип
    d - дубликаты
    l - лемматизация
    c - Подходит для категоризации

<div class="alert alert-block alert-success">
Комментарий ревьюера v. 1:
    
Ты очень детально проанализировал данные, молодец!
</div>

## Обработка пропусков

#### Пропуски в столбцах 'days_employed' и 'total_income'. 
Поскольку количество пропусков в них равно, посмотрим верно ли предположение, что это - молодеж, которая еще не работала и соответственно и не зарабатывала. Если это верно, то столбец "income_type"(пропущенные значения) - можно заполнить нулями

### столбец 'days_employed' 

In [20]:
print(data['days_employed'].value_counts())

-986.927316     1
-7026.359174    1
-4236.274243    1
-6620.396473    1
-1238.560080    1
               ..
-2849.351119    1
-5619.328204    1
-448.829898     1
-1687.038672    1
-582.538413     1
Name: days_employed, Length: 19351, dtype: int64


In [21]:
null = data.loc[(data['days_employed'].isnull()) & (data['total_income'].isnull()),'dob_years'].count()
print(f'Кол-во значений отсутствующих одновременно в обоих столбцах = {null}')
print()
print(data.loc[(data['days_employed'].isnull()) & (data['total_income'].isnull())].head(10))

Кол-во значений отсутствующих одновременно в обоих столбцах = 2174

    children  days_employed  dob_years education  education_id  \
12         0            NaN         65   среднее             1   
26         0            NaN         41   среднее             1   
29         0            NaN         63   среднее             1   
41         0            NaN         50   среднее             1   
55         0            NaN         54   среднее             1   
65         0            NaN         21   среднее             1   
67         0            NaN         52    высшее             0   
72         1            NaN         32    высшее             0   
82         2            NaN         50    высшее             0   
83         0            NaN         52   среднее             1   

            family_status  family_status_id gender  income_type  debt  \
12       гражданский брак                 1      M    пенсионер     0   
26        женат / замужем                 0      M  госслуж

Мы вывели 10 строчек с условием, что столбцы 'total_income' и 'days_employed' одновременно = "NaN"
По столбцу "dob_years" видно , что возраст варьируется от 21 до 65. Значит предположение ,что это никогда неработавшие - ложно. Во всех остальных столбцах тоже стоят разные значения. Значит будем заполнять 'total_income' средним значением (median) категории значение в каждой категории. А столбец 'days_employed' удалим.

<div class="alert alert-block alert-success">
Комментарий ревьюера v. 1:
    
Предположение оказалось неверным, Ты верно заполняешь эти пропуски медианой (несовсем верно написано "средним значением (median)", т. к. под средним значением обычно подразумевает mean). Напомню, что в зависимости от данных и задачи можно заменять пропуски средним значением, модой, какой-либо константой (например, в некоторых случаях логично пропуски заменить на 0) или использовать более сложные алгоритмы замены.
</div>

In [22]:
# Удаляем столбец 'days_employed'
data.drop('days_employed', axis='columns', inplace=True)
print(data.columns)

Index(['children', 'dob_years', 'education', 'education_id', 'family_status',
       'family_status_id', 'gender', 'income_type', 'debt', 'total_income',
       'purpose'],
      dtype='object')


### столбец 'total_income'

In [23]:
print(data.info())
print()
total_income_median = data['total_income'].median()
print(f'Медиана столбца = {total_income_median}')
print()
print(data['total_income'].describe())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 11 columns):
children            21525 non-null int64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(1), int64(5), object(5)
memory usage: 1.8+ MB
None

Медиана столбца = 145017.93753253992

count    1.935100e+04
mean     1.674223e+05
std      1.029716e+05
min      2.066726e+04
25%      1.030532e+05
50%      1.450179e+05
75%      2.034351e+05
max      2.265604e+06
Name: total_income, dtype: float64


Выбор столбца "income_type", в качестве категорий, для заполнения пропусков медианными значениями каждой категории этого столбца кажется логичным (в каждой категории более или менее данные одного порядка). Создадим income_category_list , содержащую медианы категорий.

In [24]:
income_category_list = data.groupby('income_type')['total_income'].median().sort_values()
print('Таблица медиан по категориям : ')
print(income_category_list)
print()
print('Пропущеных значений всего в total_income - ', data['total_income'].isna().sum())
print("Выведем несколько строк с пропусками:")
print(data.loc[data.loc[:,'total_income'].isna()==True].head(10))


Таблица медиан по категориям : 
income_type
в декрете           53829.130729
студент             98201.625314
пенсионер          118514.486412
безработный        131339.751676
сотрудник          142594.396847
госслужащий        150447.935283
компаньон          172357.950966
предприниматель    499163.144947
Name: total_income, dtype: float64

Пропущеных значений всего в total_income -  2174
Выведем несколько строк с пропусками:
    children  dob_years education  education_id          family_status  \
12         0         65   среднее             1       гражданский брак   
26         0         41   среднее             1        женат / замужем   
29         0         63   среднее             1  Не женат / не замужем   
41         0         50   среднее             1        женат / замужем   
55         0         54   среднее             1       гражданский брак   
65         0         21   среднее             1  Не женат / не замужем   
67         0         52    высшее             0    

In [25]:
for category in income_category_list.index:
    median = income_category_list[category]
    print(category,': Медиана = ', median)
    print(median)
    print('Пропусков было:', data[data['income_type']==category]['total_income'].isnull().sum())
    
    #data['total_income'] = data[data['income_type']==category]['total_income'].fillna(median)
    #data[data['income_type']==category]['total_income'] = data[data['income_type']==category]['total_income'].fillna(median)
    
    #data['total_income'] = data.loc[data.loc[:, 'income_type']==category, 'total_income'].fillna(median)
    data.loc[data.loc[:, 'income_type']==category, 'total_income'] = data.loc[data.loc[:, 'income_type']==category, 'total_income'].fillna(median)
    
    print('Пропусков стало:', data[data['income_type']==category]['total_income'].isnull().sum())
    print()
    
print('Осталось пропущеных значений - ', data['total_income'].isna().sum())
print()

в декрете : Медиана =  53829.13072905995
53829.13072905995
Пропусков было: 0
Пропусков стало: 0

студент : Медиана =  98201.62531401133
98201.62531401133
Пропусков было: 0
Пропусков стало: 0

пенсионер : Медиана =  118514.48641164352
118514.48641164352
Пропусков было: 413
Пропусков стало: 0

безработный : Медиана =  131339.7516762103
131339.7516762103
Пропусков было: 0
Пропусков стало: 0

сотрудник : Медиана =  142594.39684740017
142594.39684740017
Пропусков было: 1105
Пропусков стало: 0

госслужащий : Медиана =  150447.9352830068
150447.9352830068
Пропусков было: 147
Пропусков стало: 0

компаньон : Медиана =  172357.95096577113
172357.95096577113
Пропусков было: 508
Пропусков стало: 0

предприниматель : Медиана =  499163.1449470857
499163.1449470857
Пропусков было: 1
Пропусков стало: 0

Осталось пропущеных значений -  0



Перед заменой (взято из таблицы выведеной ранее)

 family_status_id gender  income_type  debt  total_income  \
12                  1      M    пенсионер     0           NaN   
26                  0      M  госслужащий     0           NaN   
29                  4      F    пенсионер     0           NaN   
41                  0      F  госслужащий     0           NaN   
55                  1      F    пенсионер     1           NaN   
65                  4      M    компаньон     0           NaN   
67                  0      F    пенсионер     0           NaN   
72                  0      M  госслужащий     0           NaN   
82                  0      F    сотрудник     0           NaN

In [26]:
print("Для проверки выведем строки где были заменены пропуски из разных категорий № 12,26,65,82")
print(data.loc[[12,26,65,82], ['total_income', 'income_type']])
print()
print('Таблица медиан по категориям : ')
print(income_category_list)
print()
print("Все медианы по категориям совпадают.")
#print(data.loc[data.loc[:,'total_income'].isna()==True].head(10))

Для проверки выведем строки где были заменены пропуски из разных категорий № 12,26,65,82
     total_income  income_type
12  118514.486412    пенсионер
26  150447.935283  госслужащий
65  172357.950966    компаньон
82  142594.396847    сотрудник

Таблица медиан по категориям : 
income_type
в декрете           53829.130729
студент             98201.625314
пенсионер          118514.486412
безработный        131339.751676
сотрудник          142594.396847
госслужащий        150447.935283
компаньон          172357.950966
предприниматель    499163.144947
Name: total_income, dtype: float64

Все медианы по категориям совпадают.


<div class="alert alert-block alert-success">
Комментарий ревьюера v. 1:
    
Отлично :)
</div>

### Столбец 'gender'

Исправим единственное значение "XNA" на "F"-женщина

In [27]:
print(data.loc[data['gender']=='XNA'])
print()
print(data['gender'].value_counts())
print()
data.loc[data['gender']=='XNA','gender'] = 'F' # Замена "XNA" на "F"
print(data['gender'].value_counts())
print()
print(data.loc[10701]) # Вывод строки где было "XNA"

       children  dob_years            education  education_id  \
10701         0         24  неоконченное высшее             2   

          family_status  family_status_id gender income_type  debt  \
10701  гражданский брак                 1    XNA   компаньон     0   

        total_income               purpose  
10701  203905.157261  покупка недвижимости  

F      14236
M       7288
XNA        1
Name: gender, dtype: int64

F    14237
M     7288
Name: gender, dtype: int64

children                               0
dob_years                             24
education            неоконченное высшее
education_id                           2
family_status           гражданский брак
family_status_id                       1
gender                                 F
income_type                    компаньон
debt                                   0
total_income                      203905
purpose             покупка недвижимости
Name: 10701, dtype: object


Замена произведена успешно.

<div class="alert alert-block alert-success">
Комментарий ревьюера v. 1:
    
Ок :)
</div>

### Столбец 'children'

Некорректные значения "20" - их 76 (20 детей - очень мало вероятно, да еще в 76 случаях, да еще нет других значений до 5 детей) и "-1" их 47(нереально). Проанализировать. По-скольку их кол-во незначительно(76+47 против 21525) возможно: 

Вариант№1 - удалить эти строки. 

Вариант№2
а) Значение "-1" могло появиться в результате написания минуса как символа'тире' (воможно по привычке), но с другой стороны возможно были бы тогда и "-2"(вполне могло быть, их 2055 по сравнению с "-1" их 4818), "-3"(могло но их значительно меньше 330) - все-таки нереально что это "тире". На клавиатуре "1" и "-" находятся делеко их одновпеменное нажатие не реально.
б) "20" вполне могло появится врезультате нажатия на NumPad 2 и 0 находятся рядом , поэтому это может значить 2 и/или 0.

Выберем, чтобы время не тратить более простой Вариант№1, проанализировав строки столбца напоследок

<div class="alert alert-block alert-success">
Комментарий ревьюера v. 1:
    
Отлично, что предлагаешь несколько вариантов замены.
</div>

In [28]:
print(data['children'].value_counts())

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64


Просмотрим в что в строках с некорректными значениями -1 и 20.

In [29]:
print(data[data['children']==-1].head(10))

      children  dob_years education  education_id          family_status  \
291         -1         46   среднее             1       гражданский брак   
705         -1         50   среднее             1        женат / замужем   
742         -1         57   среднее             1        женат / замужем   
800         -1         54   среднее             1  Не женат / не замужем   
941         -1         57   Среднее             1        женат / замужем   
1363        -1         55   СРЕДНЕЕ             1        женат / замужем   
1929        -1         38   среднее             1  Не женат / не замужем   
2073        -1         42   среднее             1              в разводе   
3814        -1         26   Среднее             1       гражданский брак   
4201        -1         41   среднее             1        женат / замужем   

      family_status_id gender  income_type  debt   total_income  \
291                  1      F    сотрудник     0  102816.346412   
705                  0      F

Закономерности не прослеживается

In [30]:
print(data[data['children']==20].head(10))

      children  dob_years education  education_id          family_status  \
606         20         21   среднее             1        женат / замужем   
720         20         44   среднее             1        женат / замужем   
1074        20         56   среднее             1        женат / замужем   
2510        20         59    высшее             0         вдовец / вдова   
2941        20          0   среднее             1        женат / замужем   
3302        20         35   среднее             1  Не женат / не замужем   
3396        20         56    высшее             0        женат / замужем   
3671        20         23   среднее             1  Не женат / не замужем   
3697        20         40   среднее             1       гражданский брак   
3735        20         26    высшее             0  Не женат / не замужем   

      family_status_id gender  income_type  debt   total_income  \
606                  0      M    компаньон     0  145334.865002   
720                  0      F

Закономерности не прослеживается, поэтому реализуем Вариант№1 - удалить эти строки. 

In [31]:
im1 = data[data['children']==-1].index # индексы 47 строк где значение -1
print(im1)
i20 = data[data['children']==20].index # индексы 76 строк где значение 20 
print(i20)
print()
#-----------Удаление--------------------------------- 
str_del_count=0
try:
    for drop_str in im1:
        data = data.drop(drop_str)
        str_del_count+=1
except:
    print('Неполучилось удалить строку')
print('Кол-во удаленных строк c "-1" =', str_del_count)

str_del_count=0
try:
    for drop_str in i20:
        data = data.drop(drop_str)
        str_del_count+=1
except:
    print('Неполучилось удалить строку')    
print('Кол-во удаленных строк c "20" =', str_del_count)
print()
print('-----------Проверка---------------------------------') 
print(data[data['children']==20].head())
print(data[data['children']==-1].head())
  
data = data.reset_index(drop=True)
print(data.info())
print('Должно остаться строк =', 21525-47-76)

#data.drop(data.loc[data['children']==20])
#print(data[data['children']==20].head())
# data = data.drop(data.loc[data['children']==-1])                          
# print(data[data['children']==-1].head())


Int64Index([  291,   705,   742,   800,   941,  1363,  1929,  2073,  3814,
             4201,  4402,  4542,  5272,  6013,  6385,  7190,  7278,  7615,
             7685,  8229,  8251,  8561,  9565,  9574, 10177, 10365, 10918,
            11109, 11268, 11617, 12304, 13786, 14357, 15147, 15435, 16129,
            16265, 17095, 17429, 17657, 18219, 19100, 19366, 19417, 20393,
            20717, 21140],
           dtype='int64')
Int64Index([  606,   720,  1074,  2510,  2941,  3302,  3396,  3671,  3697,
             3735,  3877,  5020,  5315,  5349,  5362,  6198,  6428,  7027,
             7078,  7516,  7789,  7962,  8430,  8883,  8909,  8950,  9889,
            10194, 10377, 10736, 10782, 11449, 11489, 11715, 11937, 12212,
            12226, 12319, 12390, 12534, 12909, 12910, 13256, 13291, 13471,
            13489, 13618, 14355, 14474, 14782, 15313, 15674, 15762, 15812,
            15976, 16371, 16795, 16886, 17093, 17134, 17286, 17509, 18476,
            18493, 18756, 18890, 19271, 19562, 

<div class="alert alert-block alert-success">
Комментарий ревьюера v. 1:
    
Молодец, что попробовал выявить закономерности в данных.
</div>

### Столбец 'dob_years'

101 некорректное значение. Проверить в каких строках оно встречается, и если на результат не будет оказывать влияния, то можно их удалить (их незначительное кол-во 101 против 21525 - меньше 0,5%)

In [32]:
print(data['dob_years'].value_counts().sort_index())

0     100
19     14
20     51
21    110
22    183
23    252
24    263
25    356
26    406
27    490
28    501
29    543
30    536
31    556
32    506
33    577
34    597
35    614
36    553
37    531
38    595
39    572
40    603
41    603
42    592
43    510
44    543
45    494
46    469
47    480
48    536
49    505
50    509
51    446
52    483
53    457
54    476
55    441
56    482
57    457
58    461
59    441
60    376
61    353
62    351
63    268
64    263
65    194
66    183
67    167
68     99
69     83
70     65
71     58
72     33
73      8
74      6
75      1
Name: dob_years, dtype: int64


In [33]:
print(data[data['dob_years']==0].head(15))

      children  dob_years education  education_id          family_status  \
99           0          0   Среднее             1        женат / замужем   
149          0          0   среднее             1              в разводе   
270          3          0   среднее             1        женат / замужем   
577          0          0   среднее             1        женат / замужем   
1033         0          0    высшее             0              в разводе   
1141         0          0   среднее             1        женат / замужем   
1167         0          0   среднее             1        женат / замужем   
1377         0          0    высшее             0        женат / замужем   
1881         0          0    высшее             0  Не женат / не замужем   
1889         0          0   среднее             1         вдовец / вдова   
2071         0          0   среднее             1        женат / замужем   
2273         0          0   среднее             1         вдовец / вдова   
2458        

Закономерности не прослеживается, поэтому удалим эти 100 (было 101) строк.

In [34]:
i0 = data[data['dob_years']==0].index # индексы 100 строк где значение 0 
print(i0, 'Кол-во =', len(i0))
print()
#-----------Удаление--------------------------------- 
str_del_count=0
for drop_str in i0:
    data = data.drop(drop_str)
    str_del_count+=1
print('Кол-во удаленных строк c "0" =', str_del_count)
#-----------Проверка---------------------------------  
print()
print(data[data['dob_years']==0].head())
print()  
data = data.reset_index(drop=True)
print()  
print(data.info())
print('Должно остаться строк =', 21402-100)

Int64Index([   99,   149,   270,   577,  1033,  1141,  1167,  1377,  1881,
             1889,  2071,  2273,  2458,  2476,  2858,  3031,  3134,  3687,
             4044,  4127,  4899,  4907,  4991,  6042,  6376,  6380,  6638,
             6746,  6799,  6816,  6827,  7001,  7040,  7042,  7217,  7308,
             8020,  8373,  8529,  8627,  8739,  9014,  9103,  9538, 10112,
            10136, 10253, 10311, 10329, 10490, 10540, 11019, 11229, 11407,
            11420, 11514, 11601, 11925, 11997, 12159, 12230, 12333, 12658,
            13044, 13114, 13159, 13338, 13348, 13364, 13444, 13663, 13889,
            14395, 14432, 14526, 14577, 15057, 15211, 15324, 15348, 15797,
            15802, 15952, 16767, 16827, 17117, 17207, 17512, 18074, 18312,
            18434, 18627, 18745, 19008, 19261, 19716, 20346, 20461, 21060,
            21194],
           dtype='int64') Кол-во = 100

Кол-во удаленных строк c "0" = 100

Empty DataFrame
Columns: [children, dob_years, education, education_id, family_

С пропусками и удалениями закончили

### Вывод

In [35]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21302 entries, 0 to 21301
Data columns (total 11 columns):
children            21302 non-null int64
dob_years           21302 non-null int64
education           21302 non-null object
education_id        21302 non-null int64
family_status       21302 non-null object
family_status_id    21302 non-null int64
gender              21302 non-null object
income_type         21302 non-null object
debt                21302 non-null int64
total_income        21302 non-null float64
purpose             21302 non-null object
dtypes: float64(1), int64(5), object(5)
memory usage: 1.8+ MB


##### Итого у нас было 21525 строк и 12 столбцов , стало 11 столбцов и  21302 строки - на 1% меньше, - вполне допустимо.

<div class="alert alert-block alert-success">
Комментарий ревьюера v. 1:
    
Да, процент удаляемых данных небольшой (меньше 5%), поэтому можно удалить.
</div>

## Замена типа данных

Заменим тип в 2 столбцац: 1) "debt" на bool; 2)"total_income" на int

In [36]:
print('debt - перед изменением: ',data[data['debt']!=0]['debt'].head())
data['debt'] = data['debt'].astype(bool)
print('debt - после изменения: ', data[data['debt']==1]['debt'].head())

print()
print('total_income - перед изменением: ', data['total_income'].head())
data['total_income'] = data['total_income'].astype(int)
print('total_income - после изменения: ', data['total_income'].head())

data.info()

debt - перед изменением:  14    1
32    1
38    1
55    1
75    1
Name: debt, dtype: int64
debt - после изменения:  14    True
32    True
38    True
55    True
75    True
Name: debt, dtype: bool

total_income - перед изменением:  0    253875.639453
1    112080.014102
2    145885.952297
3    267628.550329
4    158616.077870
Name: total_income, dtype: float64
total_income - после изменения:  0    253875
1    112080
2    145885
3    267628
4    158616
Name: total_income, dtype: int64
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21302 entries, 0 to 21301
Data columns (total 11 columns):
children            21302 non-null int64
dob_years           21302 non-null int64
education           21302 non-null object
education_id        21302 non-null int64
family_status       21302 non-null object
family_status_id    21302 non-null int64
gender              21302 non-null object
income_type         21302 non-null object
debt                21302 non-null bool
total_income        21302 non-nul

### Вывод

Поменяли успешно тип в обоих столбцах. Теперь всё будет красивее, понятней и быстрее :) Выбор метода .astype продиктован простотой и эффективностью применения.

<div class="alert alert-block alert-success">
Комментарий ревьюера v. 1:
    
Теперь с типами данных всё в порядке :)
</div>

## Обработка дубликатов

### Столбец 'education'

Выведем информацию по столбцам с образованием методом .value_counts()

In [37]:
print(data['education'].value_counts(), data['education_id'].value_counts())

среднее                13609
высшее                  4666
СРЕДНЕЕ                  764
Среднее                  700
неоконченное высшее      663
ВЫСШЕЕ                   270
Высшее                   266
начальное                250
Неоконченное высшее       47
НЕОКОНЧЕННОЕ ВЫСШЕЕ       29
НАЧАЛЬНОЕ                 17
Начальное                 15
ученая степень             4
Ученая степень             1
УЧЕНАЯ СТЕПЕНЬ             1
Name: education, dtype: int64 1    15073
0     5202
2      739
3      282
4        6
Name: education_id, dtype: int64


Можно предположить(с большой вероятностью), что причиной возникновение дубликатов стало то, что таблицу заполняли 3 разные человека(или групп людей), так как каждая категория четко написана в 3 вариантах 1) Большими 2) Маленькими 3) С большой.

Видно что в 'education' стоят одни и теже значения, но в разном регистре. Поэтому попробуем убрать дубликаты сделав все значения столбца 'education' прописными (в нижнем регистре). И потом проверим совпадают ли кол-во значений по категориям в 'education' со столбцом "education_id".

In [38]:
data['education'] = data['education'].str.lower()
print(data['education'].value_counts(), data['education_id'].value_counts())

среднее                15073
высшее                  5202
неоконченное высшее      739
начальное                282
ученая степень             6
Name: education, dtype: int64 1    15073
0     5202
2      739
3      282
4        6
Name: education_id, dtype: int64


### Вывод

Дубликаты были обнаружены в столбцах 'education' и 'purpose'. В столбце 'education' - дубликаты убрали изменив регистр названия на нижний, даже не потребовалось ничего удалять и перезаписывать значения. 

Столбец 'purpose' - будем обрабатывать в следующем разделе Лемматизация.

<div class="alert alert-block alert-success">
Комментарий ревьюера v. 1:
    
Молодец, что исследовал дубликаты.
</div>

## Лемматизация

### Столбец 'purpose'

Нужно совпадающие цели привезти к одному названию

Для этого создадим DataFrame с 3 столбцами Строка_Оригинал/Строка_Лемматизировання/Лемма. Затем создадим словарь Лема-ЦелевоеЗначение, и применим замену по этому словарю в столбце 'purpose' ко всему DataFrame data.

In [39]:
print(data['purpose'].value_counts())

свадьба                                   791
на проведение свадьбы                     768
сыграть свадьбу                           764
операции с недвижимостью                  670
покупка коммерческой недвижимости         658
покупка жилья для сдачи                   649
операции с коммерческой недвижимостью     644
операции с жильем                         642
покупка жилья для семьи                   639
жилье                                     636
покупка жилья                             635
недвижимость                              628
операции со своей недвижимостью           626
строительство собственной недвижимости    626
строительство недвижимости                620
строительство жилой недвижимости          619
покупка своего жилья                      618
покупка недвижимости                      616
ремонт жилью                              607
покупка жилой недвижимости                600
на покупку своего автомобиля              501
заняться высшим образованием      

Закономерности в возникновении дубликатов не прослеживается. Причиной возникновения дубликатов скорее всего стало то, что данные в таблице запонялись большим количеством людей и каждый заполнял по своему.

<div class="alert alert-block alert-success">
Комментарий ревьюера v. 1:
    
Вполне логичное предположение :)
</div>

In [40]:
from pymystem3 import Mystem
m = Mystem()
from collections import Counter

columns = ['original', 'lemms', 'good']
purpose_list = pd.DataFrame(data=[],columns=columns)
purpose_list['original'] = data['purpose'].value_counts().index

for i in purpose_list.index:
    purpose_list.loc[i,'lemms'] = ' '.join(m.lemmatize(purpose_list.loc[i, 'original']))
    #print(m.lemmatize(purpose_list.loc[i, 'original']), ''.join(m.lemmatize(purpose_list.loc[i, 'original'])))
print(purpose_list)

                                  original  \
0                                  свадьба   
1                    на проведение свадьбы   
2                          сыграть свадьбу   
3                 операции с недвижимостью   
4        покупка коммерческой недвижимости   
5                  покупка жилья для сдачи   
6    операции с коммерческой недвижимостью   
7                        операции с жильем   
8                  покупка жилья для семьи   
9                                    жилье   
10                           покупка жилья   
11                            недвижимость   
12         операции со своей недвижимостью   
13  строительство собственной недвижимости   
14              строительство недвижимости   
15        строительство жилой недвижимости   
16                    покупка своего жилья   
17                    покупка недвижимости   
18                            ремонт жилью   
19              покупка жилой недвижимости   
20            на покупку своего ав

Из списка целей кредита видно, что одна и таже цель записаны по разному. 
Выделим цели и леммы и составим словарь их соответствия Лемма - Цель:

свадьба - свадьба

жилье,недвижимость - недвижимость

автомобиль - автомобиль

образование - образование

<div class="alert alert-block alert-success">
Комментарий ревьюера v. 1:
    
Отлично, цели кредита верно выделены.
</div>

In [41]:
dict_lem_purpose = dict({'свадьба':'свадьба', 
                    'жилье':'недвижимость', 
                    'недвижимость':'недвижимость',
                    'автомобиль':'автомобиль', 
                    'образование':'образование'})

#-------------Вариант с .apply--------------------------------
# Напишем функцию,которая принимает значение столбца (строка из "purpose") и меняет его на то которое в словаре (dict_lem_purpose)
def make_good_category_str(row):
    str_lem_list = m.lemmatize(row)
    for lem in str_lem_list:
        if lem in dict_lem_purpose.keys():
            #print(lem, dict_lem_purpose[lem], str_lem_list)
            return dict_lem_purpose[lem]
# Произведем замену дубликатов с помощью нашей функции "make_good_category_str"
purpose_list['good'] = purpose_list['original'].apply(make_good_category_str)        
#------------------------------------------------------------- 

#-------------Старый Вариант с циклом-------------------------
# for i in purpose_list.index:    
#     str_lem_list = m.lemmatize(purpose_list.loc[i, 'original'])
#     #print(i, str_lem_list)
#     for lem in str_lem_list:
#         if lem in dict_lem_purpose.keys():
#             #print(lem, dict_lem_purpose[lem], str_lem_list)
#             purpose_list.loc[i, 'good'] = dict_lem_purpose[lem]
#             break
#-------------------------------------------------------------        
    
print(purpose_list)

                                  original  \
0                                  свадьба   
1                    на проведение свадьбы   
2                          сыграть свадьбу   
3                 операции с недвижимостью   
4        покупка коммерческой недвижимости   
5                  покупка жилья для сдачи   
6    операции с коммерческой недвижимостью   
7                        операции с жильем   
8                  покупка жилья для семьи   
9                                    жилье   
10                           покупка жилья   
11                            недвижимость   
12         операции со своей недвижимостью   
13  строительство собственной недвижимости   
14              строительство недвижимости   
15        строительство жилой недвижимости   
16                    покупка своего жилья   
17                    покупка недвижимости   
18                            ремонт жилью   
19              покупка жилой недвижимости   
20            на покупку своего ав

<div class="alert alert-block alert-success">
Комментарий ревьюера v. 1:
    
Обработка целей кредита проведена верно, молодец!
</div>

Убрать дубликаты получилось. Теперь применим замену ко всему датасету "data" - к столбцу "purpose" и запишем результат в новый столбец 'comm_purpose'

In [42]:
#-------------Старый Вариант с циклом----------------------
# for i in data.index:    
#     str_lem_list = m.lemmatize(data.loc[i, 'purpose'])
#     #print(i, str_lem_list)
#     for lem in str_lem_list:
#         if lem in dict_lem_purpose.keys():
#             #print(lem, dict_lem_purpose[lem], str_lem_list)
#             data.loc[i, 'purpose'] = dict_lem_purpose[lem]
#             break
#------------------------------------------------------------

#-------------Вариант с .apply--------------------------------
data['comm_purpose'] = data['purpose'].apply(make_good_category_str)
#------------------------------------------------------------

"Старый Вариант с циклом" - обрабатывал ДатаСет около 30 секунд ! - довольно долго( даже подумал, что всё подвисло:) )
Поэтому было решено применить метод .apply. В результате обработка за 2-3 секнды ! Выбор очевиден. (В принципе можно было еще еще Вариант№3(он бы работал еще быстрее) - создать словарь из DataFrame "purpose_list", заполнив его так: ключ - 1 столбец, а значения из 3 столбца; и использовать его для замены в главном Датасете - но в следующий раз, когда будет время) 

<div class="alert alert-block alert-success">
Комментарий ревьюера v. 1:
    
Отличная идея использовать метод .apply :)
    
Здорово, что усовершенствовал первоначальный вариант.
</div>

In [43]:
#-------------Вариант со словарем dict_data_purpose--------------------------------
# for i in purpose_list.index:
#     dict_data_purpose = {purpose_list.loc[i, 'original']:purpose_list.loc[i, 'good']} # Как создать словарь из DataFrame???
#     print(i, purpose_list.loc[i, 'good'], purpose_list.loc[i, 'original'])
# print(dict_data_purpose)

In [44]:
print(data['comm_purpose'].value_counts())
print()
print(data.head())

недвижимость    10733
автомобиль       4267
образование      3979
свадьба          2323
Name: comm_purpose, dtype: int64

   children  dob_years education  education_id     family_status  \
0         1         42    высшее             0   женат / замужем   
1         1         36   среднее             1   женат / замужем   
2         0         33   среднее             1   женат / замужем   
3         3         32   среднее             1   женат / замужем   
4         0         53   среднее             1  гражданский брак   

   family_status_id gender income_type   debt  total_income  \
0                 0      F   сотрудник  False        253875   
1                 0      F   сотрудник  False        112080   
2                 0      M   сотрудник  False        145885   
3                 0      M   сотрудник  False        267628   
4                 1      F   пенсионер  False        158616   

                      purpose  comm_purpose  
0               покупка жилья  недвижимость 

### Вывод

Лемматизацию провели успешно. Теперь у нас 4 цели кредита вместо 37.

недвижимость - 10733 / автомобиль - 4267 / образование - 3979 / свадьба - 2323 

<div class="alert alert-block alert-success">
Комментарий ревьюера v. 1:
    
Отлично, теперь у нас всего 4 цели для кредита.
</div>

## Категоризация данных

Для категоризации подходят два целочисленных столбца "dob_years" и "total_income" так как там много разных значений - поэтому нам надо объединить их в категории. В остальных столбцах к категориям уже все приведено в результате преобразования ДатаСета(education, children, gender, income_type, purpose) и создан новый столбец "comm_purpose" (обобщенные цели кредита) или категории уже определены и этого не требуется (education_id, family_status, family_status_id, debt). 

В столбце "dob_years" - категоризацию проводить не будем , так как для ответа на поставленные вопросы этот столбец не пригодится.

Итак столбец "total_income". Выведем распределение его значений с помощюего describe() . 

In [45]:
print(data['total_income'].describe())

count    2.130200e+04
mean     1.652897e+05
std      9.827766e+04
min      2.066700e+04
25%      1.077762e+05
50%      1.425940e+05
75%      1.955475e+05
max      2.265604e+06
Name: total_income, dtype: float64


Чтобы разбить на категории нужно более подробный анализ - оценим распределение значений с помощью процентиля.

In [46]:
print()
for proc in range(2,100,3):
    quant = int(data['total_income'].quantile(proc/100))
    print(f'Процентиль {proc}% = {quant}')



Процентиль 2% = 51449
Процентиль 5% = 64748
Процентиль 8% = 73294
Процентиль 11% = 80812
Процентиль 14% = 86834
Процентиль 17% = 92963
Процентиль 20% = 98619
Процентиль 23% = 103957
Процентиль 26% = 109611
Процентиль 29% = 114481
Процентиль 32% = 118514
Процентиль 35% = 122555
Процентиль 38% = 128500
Процентиль 41% = 134023
Процентиль 44% = 139569
Процентиль 47% = 142594
Процентиль 50% = 142594
Процентиль 53% = 146982
Процентиль 56% = 152411
Процентиль 59% = 159001
Процентиль 62% = 165350
Процентиль 65% = 172287
Процентиль 68% = 173786
Процентиль 71% = 182610
Процентиль 74% = 192235
Процентиль 77% = 201766
Процентиль 80% = 214169
Процентиль 83% = 227338
Процентиль 86% = 243551
Процентиль 89% = 262211
Процентиль 92% = 286843
Процентиль 95% = 331802
Процентиль 98% = 422210


<div class="alert alert-block alert-success">
Комментарий ревьюера v. 1:
    
Молодец, что основательно подходишь к категоризации :)
    
Использовать перцентили для определения границ категорий отличная идея, здесь может помочь построение гистограммы и "ящика с усами" (boxplot).
</div>

Создадим новые столбцы  "income_rating"(типа Object) и "income_rating_id"(типа int) (для удобства). Итак доход разобьем на 6 категорий (от 4 до 7 категорий допустимые варианты, ниже 4 уже не очень информативно, а выше 7 уже запутаться можно, пусть будет 6 категорий):

сверхмалый / 1 - до 55000

малый / 2 - до 100000

средний / 3 - до 190000

сверхсредний / 4 - до 240000

большой / 5 - до 420000

сверхбольшой / 6 - выше 420000

In [47]:
data['income_rating'] = 'NaN'
data['income_rating_id'] = 'NaN'
data.loc[data['total_income'] < 55000, ['income_rating', 'income_rating_id']] = ['сверхмалый', 1]
data.loc[(55000 <= data['total_income']) & ( data['total_income'] < 100000), ['income_rating', 'income_rating_id']] = ['малый', 2]
data.loc[(100000 <= data['total_income']) & ( data['total_income'] < 190000), ['income_rating', 'income_rating_id']] = ['средний', 3]
data.loc[(190000 <= data['total_income']) & ( data['total_income'] < 240000), ['income_rating', 'income_rating_id']] = ['сверхсредний', 4]
data.loc[(240000 <= data['total_income']) & ( data['total_income'] < 420000), ['income_rating', 'income_rating_id']] = ['большой', 5]
data.loc[data['total_income'] >= 420000, ['income_rating', 'income_rating_id']] = ['сверхбольшой', 6]
print(data['income_rating'].value_counts())
print(data['income_rating_id'].value_counts())
print(data.head())
print(data.info())

средний         11189
малый            3849
большой          2662
сверхсредний     2595
сверхмалый        572
сверхбольшой      435
Name: income_rating, dtype: int64
3    11189
2     3849
5     2662
4     2595
1      572
6      435
Name: income_rating_id, dtype: int64
   children  dob_years education  education_id     family_status  \
0         1         42    высшее             0   женат / замужем   
1         1         36   среднее             1   женат / замужем   
2         0         33   среднее             1   женат / замужем   
3         3         32   среднее             1   женат / замужем   
4         0         53   среднее             1  гражданский брак   

   family_status_id gender income_type   debt  total_income  \
0                 0      F   сотрудник  False        253875   
1                 0      F   сотрудник  False        112080   
2                 0      M   сотрудник  False        145885   
3                 0      M   сотрудник  False        267628   
4      

### Вывод

Разбили доходы на 6 категорий по возрастанию, при этом добавили 2 столбца "income_rating" - название категории и "income_rating_id" - её id-номер (чтобы удобней было пользоваться - сортировать, да и нагляднее так)

<div class="alert alert-block alert-success">
Комментарий ревьюера v. 1:
    
Отлично, категоризация дохода проведена верно.
</div>

# Шаг 3. Ответьте на вопросы

- Есть ли зависимость между наличием детей и возвратом кредита в срок?

Чтобы увидеть и проанализировать есть ли зависимость - составим сводную таблицу методом .pivot_table. Необходимые столбцы - 'children' и 'debt'. Для группировки логично будет взять столбцы, связанные с семьей и детьми т.е. 'family_status_id' и 'family_status' и для интереса еще добавим 'gender'.


<div class="alert alert-block alert-success">
Комментарий ревьюера v. 1:
    
Здорово, что используешь .pivot_table.
</div>

In [48]:
data_pivot_children = data.pivot_table(index=['family_status_id', 'family_status', 'gender'], columns='children', values='debt', aggfunc='sum')
print(data_pivot_children)

children                                           0      1     2     3    4  \
family_status_id family_status         gender                                  
0                женат / замужем       F       290.0  127.0  96.0  10.0  0.0   
                                       M       224.0  117.0  49.0   7.0  3.0   
1                гражданский брак      F       131.0   69.0  24.0   7.0  0.0   
                                       M        96.0   49.0   6.0   1.0  0.0   
2                вдовец / вдова        F        42.0    6.0   3.0   0.0  0.0   
                                       M        10.0    1.0   0.0   NaN  NaN   
3                в разводе             F        44.0   13.0   4.0   0.0  0.0   
                                       M        11.0    8.0   3.0   1.0  NaN   
4                Не женат / не замужем F        82.0   28.0   7.0   0.0  1.0   
                                       M       128.0   23.0   2.0   1.0  NaN   

children                               

Неплохо, но для краткости и большей информативности уберем группировку по 'family_status_id', 'family_status' и сотавим только 'gender'

In [49]:
data_pivot_children = data.pivot_table(index=['gender'], columns='children', values='debt', aggfunc='sum')
print(data_pivot_children)

children      0      1      2     3    4    5
gender                                       
F         589.0  243.0  134.0  17.0  1.0  0.0
M         469.0  198.0   60.0  10.0  3.0  0.0


Напишем функцию которая принимает список столбцов (или 1 столбец) для анализа просрочки и выводит таблицу с колонками: принятые столбцы / а) кол-во невыплат / б) кол-во записей в категории / доля невыплат по категории - а) делим на б)

! ВАРИАНТ С ФУНКЦИЕЙ БЫЛ ТРЕТЬИМ ПОЭТОМУ ПОРЯДОК НЕМНОГО СБИТ, а предыдущие варианты закомментированы. Этот вариант был выбран потому, что он отражает долю просрочек в категории, а не только их количество.

In [50]:
def analyse_debt(column_list):
    analyse_table = data.groupby(column_list).agg({'debt':['sum', 'count']})
    analyse_table[('debt', 'part/cat,%')] = analyse_table[('debt', 'sum')] / analyse_table[('debt', 'count')]*100
    return analyse_table

print(analyse_debt(['children']))

            debt                  
             sum  count part/cat,%
children                          
0         1058.0  14080   7.514205
1          441.0   4802   9.183673
2          194.0   2042   9.500490
3           27.0    328   8.231707
4            4.0     41   9.756098
5            0.0      9   0.000000


In [51]:
# data_grouped = data_final.groupby(['category_name','subcategory_name','source']).agg({'visits':'sum'})
# print(data_grouped.head(10))

# data_grouped = data_final.groupby(['children','debt']).agg({'visits':'sum'})
# print(data_grouped.head(10))

### Вывод

Прекрасно! Из полученных данных четко прослеживается линейная зависимость, что чем больше детей тем выше платежная дисциплина! Как ни странно люди у которых нет детей имеют больше просроченных кредитов - вроде бы и детей нет, затрат должно быть меньше, а платят хуже. Похоже, что у кого детей больше - у них мотивация выше и их жизнь заставляет "крутится" лучше:). Также можно отметить, что женщины допускают просрочку на 25-35% чаще чем мужчины.

С дугой стороны доля невыплат меньше всего у тех укого нет детей - 7,5%. А больше всего доля просрочки у тех у кого 1,2 и как на странно 4 ребенка. У кого 3 ребенка - с них начинается многодетная семья, наверное для них существуют разные преференции (со стороны государства), поддерживающие их финансовое состоянии. А когда появляется 4 ребенок на всех уже финансов не хватает. 

И еще можно заметить то, что у кого 2 ребенка высока доля просрочки - 9,5% (почти как и в ситуации с 4 детьми), хотя по поддержке государства за вторго ребенка семьи получают материнский капиталл. Можно сделать вывод, что не очень то он и помогает.

<div class="alert alert-block alert-success">
Комментарий ревьюера v. 1:
    
Всё верно, отличный анализ. Ещё можно использовать корреляционную функцию для выявления линейной зависимости между переменными.
</div>

- Есть ли зависимость между семейным положением и возвратом кредита в срок?

--- Старый, простой вариант ----  Сгруппируем по столбцам 'family_status_id', 'family_status' и посчитаем должников (в столбце 'debt').

In [52]:
#------Старый простой вариант------------------------
# data_pivot_family_status = data.groupby(['family_status_id', 'family_status'])['debt'].sum()
# print(data_pivot_family_status)

Создадим таблицу с 5 колонками. Как и в первом варианте первые 3 колонки "Сгруппируем по столбцам 'family_status_id', 'family_status' и посчитаем должников (в столбце 'debt' метод sum)". Добавим 4 колонку - с количеством всех записей по каждой категории; и 5 колона - это 3колонка/4колонка в процентах - это показатель долей просрочки в каждой категории.

In [53]:
#------Вариант2 с долей просрочки в каждой категории--------------
#------Создадим таблицу с 5 колонками---------------------------- 

data_pivot_family_status = data.groupby(['family_status_id', 'family_status']).agg({'debt':['sum', 'count']})

# print(data_pivot_family_status)
# print(data_pivot_family_status.columns)
# print(data_pivot_family_status.index)
# print(data_pivot_family_status.values)

data_pivot_family_status[('debt', 'part,%')] = data_pivot_family_status[('debt', 'sum')] / data_pivot_family_status[('debt', 'count')]*100
print(data_pivot_family_status)

                                         debt                 
                                          sum  count    part,%
family_status_id family_status                                
0                женат / замужем        923.0  12254  7.532234
1                гражданский брак       383.0   4139  9.253443
2                вдовец / вдова          62.0    947  6.546990
3                в разводе               84.0   1179  7.124682
4                Не женат / не замужем  272.0   2783  9.773626


In [54]:
#----Вариант3----
print(analyse_debt(['family_status_id', 'family_status']))

                                         debt                  
                                          sum  count part/cat,%
family_status_id family_status                                 
0                женат / замужем        923.0  12254   7.532234
1                гражданский брак       383.0   4139   9.253443
2                вдовец / вдова          62.0    947   6.546990
3                в разводе               84.0   1179   7.124682
4                Не женат / не замужем  272.0   2783   9.773626


### Вывод

Как ни странно больше всего просрочек у женатых/замужных , казалось бы живут вдвоем, скорее всего и работают оба, а тем не менее и просрочка выше! На втором месте гражданский брак - тоже живут вместе(после женатых теперь просрочки кажутся логичными и закономерными, наверное здесь большая доля тех кто еще не поженился). И на третьем месте одинокие - тоже довольно много. Меньше всего просрочек у тех кто каким-либо образом расстался со своим(ей) супругой(ом). Как видно из 5 столбца доля просрочки (около 9,5%) выше у тех кто не женат/замужем или живут гражданским браком.

<div class="alert alert-block alert-success">
Комментарий ревьюера v. 1:
    
Действительно интересные результаты.
</div>

- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

Здесь нам помогут новые столбцы, которые мы создали при категоризации: 'income_rating_id' и 'income_rating'

In [55]:
#------Старый простой вариант------------------------
# data_pivot_income_rating = data.groupby(['income_rating_id', 'income_rating'])['debt'].sum()
# print(data_pivot_income_rating)

Создадим таблицу по аналогии с предыдущим заданием (3столбец - кол-во невыплат / 4 - кол-во записей в категории / 5 - показатель долей просрочки в каждой категории т.е. 3ст/4ст в %)

In [56]:
#------Вариант с долей просрочки в каждой категории--------------
# data_pivot_income_rating = data.groupby(['income_rating_id', 'income_rating']).agg({'debt':['sum', 'count']})
# data_pivot_income_rating[('debt', 'sum/count, %')] = data_pivot_income_rating[('debt', 'sum')] / data_pivot_income_rating[('debt', 'count')]*100
# print(data_pivot_income_rating)

#--Тоже самое через функцию analyse_debt из предыдущего вопроса
print(analyse_debt(['income_rating_id', 'income_rating']))

                                 debt                  
                                  sum  count part/cat,%
income_rating_id income_rating                         
1                сверхмалый      35.0    572   6.118881
2                малый          318.0   3849   8.261886
3                средний        956.0  11189   8.544106
4                сверхсредний   199.0   2595   7.668593
5                большой        192.0   2662   7.212622
6                сверхбольшой    24.0    435   5.517241


### Вывод

По столбцу sum видно, что зависимость невыплат от дохода прослеживается по "нормальному распределению Гаусса" - по краям групп дохода (это те у кого самые большие-сверхбольшой и самые маленькие-сверхмалый доходы) кол-во просрочек маленькое и просрочки возрастают с приближением к середине, и пик ярко выражен в средний категории дохода. Доля невыплат похожа на нормальное распределение, только пик просрочек смещен в категории средний-малый

<div class="alert alert-block alert-success">
Комментарий ревьюера v. 1:
    
Интересное наблюдение, что "зависимость невыплат от дохода прослеживается по "нормальному распределению Гаусса"". 
</div>

- Как разные цели кредита влияют на его возврат в срок?

    Воспользуемся методом из двух предыдущих вопросов (эх, давно уже надо было написать функцию.    --- p.s.-написал)

In [57]:
#------Вариант с долей просрочки в каждой категории--------------
# data_pivot_purpose = data.groupby(['comm_purpose']).agg({'debt':['sum', 'count']})
# data_pivot_purpose[('debt', 'part,%')] = data_pivot_purpose[('debt', 'sum')] / data_pivot_purpose[('debt', 'count')]*100
# print(data_pivot_purpose)

#--Тоже самое через функцию analyse_debt из предыдущего вопроса
print(analyse_debt(['comm_purpose']))

               debt                  
                sum  count part/cat,%
comm_purpose                         
автомобиль    397.0   4267   9.303961
недвижимость  777.0  10733   7.239355
образование   369.0   3979   9.273687
свадьба       181.0   2323   7.791649


### Вывод

Больше всего просроченых кредитов в категории "недвижимость" (скорее всего это ипотека и вообще эта категория размером с оставшиеся 3 - 10733 записи - это самый долгосрочный кредит), но зато доля просрочек по категории 7,24% - самая низкая. 

Категории "автомобиль" и "образование" находятся посередине по кол-ву просрочек, а доля невыплат здесь значительно выше - около 9,3%. 

И меньше всего просрочек соответственно в более мелкой категории "свадьба" и доля просрочек 7,8% (на 0,5% выше чем "недвижимость") - невысоко по сравнению "автомобиль" и "образование". 

<div class="alert alert-block alert-success">
Комментарий ревьюера v. 1:
    
Ты верно определил влияние цели кредита на просрочку платежей. Хотя на самом деле разница незначительная (несколько процентов), поэтому спорный вопрос являются ли найденные закономерности статистически значимыми.
</div>

# Шаг 4. Общий вывод

В общем поставленная задача выполнена. В ходе процесса мы проанализировали данные. При предобработке данных было удалено около 1% некорректных строк и 1 некорректный столбец(который для поставленных задач не был необходим). Поменяли в двух столбцах типы - для лучшей наглядности и более быстрой обработки. Удалили дубликаты, применяли лемматизацию. Для одного столбца сделали категоризацию. После подготовки данных, сделали сводные таблицы (ипользовали функцию) и ответили на все 4 поставленных вопроса.  Были выявили закономерности в просрочках по кредитам(где-то ярко выраженные, а где-то нет).

<div class="alert alert-block alert-warning">
Комментарий ревьюера v. 1:
    
Ты провёл очень детальный анализ данных, отличная работа. Но было бы здорово добавить краткое описание закономерностей закономерности в просрочках по кредитам.
</div>

### Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  открыт файл;
- [x]  файл изучен;
- [x]  определены пропущенные значения;
- [x]  заполнены пропущенные значения;
- [x]  есть пояснение, какие пропущенные значения обнаружены;
- [x]  описаны возможные причины появления пропусков в данных;
- [x]  объяснено, по какому принципу заполнены пропуски;
- [x]  заменен вещественный тип данных на целочисленный;
- [x]  есть пояснение, какой метод используется для изменения типа данных и почему;
- [x]  удалены дубликаты;
- [x]  есть пояснение, какой метод используется для поиска и удаления дубликатов;
- [x]  описаны возможные причины появления дубликатов в данных;
- [x]  выделены леммы в значениях столбца с целями получения кредита;
- [x]  описан процесс лемматизации;
- [x]  данные категоризированы;
- [x]  есть объяснение принципа категоризации данных;
- [x]  есть ответ на вопрос: "Есть ли зависимость между наличием детей и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между семейным положением и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между уровнем дохода и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Как разные цели кредита влияют на его возврат в срок?";
- [x]  в каждом этапе есть выводы;
- [x]  есть общий вывод.